Build directory to gitignore & composer udpate

madmages opened the pull request:
Build directory added to gitignore for misscommit case
& composer update

maddmages-l2lnch committed:
Build directory to gitignore

maddmages-l2lnch committed:
composer.json & lock update

maddmages-l2lnch committed:
Massive impromevents

maddmages-l2lnch committed:
strict type of methods return

bayne reviewed (changes requested):
Really some excellent stuff here thanks for the contribution!

One minor thing (I hacked together a PHP 7.2 compatible version of this a while back but never committed it), can we get rid of `Object.php.twig` and all references to `Object` and just call it `Type` instead? 

Looks like `object` that is now a reserved keyword https://www.php.net/manual/en/reserved.other-reserved-words.php

maddmages-l2lnch committed:
Renamed Object to Type

madmages commented:
Renamed object to type.
Added full class name to shema

maddmages-l2lnch committed:
Massive changes

madmages commented:
also
+ Changed schema
+ Added type description (before class declaration). 
+ Removed link from property to class phpdoc
+ a bit precise type detection

maddmages-l2lnch committed:
fix first parameter of method

madmages commented:
+ fixed: first parameter of all methods was omitted

bayne commented on src/Command/GenerateClientCommand.php:
```diff
@@ -2,86 +2,99 @@
+                'TypedClientInterface',
+                ['namespace' => self::BASE_NAMESPACE, 'schema' => $schema],
+            ],
+        ];
```
Why two different `ClientInterface`s? If you have a use for it, I'm cool with keeping it. But otherwise let's only have the TypedClientInterface

bayne commented on src/Command/GenerateSchemaCommand.php:
```diff
@@ -57,98 +83,91 @@ protected function execute(InputInterface $input, OutputInterface $output)
-                            'fields' => $fields
+                        $this->schema['types'][$methodOrObjectName] = [
+                            'name'         => $methodOrObjectName,
+                            'link'         => $node->filter('a')->attr('href'),
```
```suggestion
                            'link'         => self::URL.$node->filter('a')->attr('href'),
```
The generated output was missing the full URL

madmages replied:
i got bot api html from https://core.telegram.org/bots/api
1. Open in browser
2. Save to schema.html
3. Parse
and i got full url

```
namespace Bayne\Telegram\Bot\Type;

/**
 * https://core.telegram.org/bots/api#chat
 *
 * This object represents a chat. 
 */
class Chat extends AbstractType
{

bayne commented on templates/ClientInterface.php.twig:
```diff
@@ -14,25 +13,20 @@ interface ClientInterface {
 
-{% else %}
-        ${{ parameter.name|camelize }}{% if not parameter.required %} = null{% endif %}{% if loop.index < method.parameters|length %},{% endif %}
+        {{ renderType(parameter.type, namespace, false) }} ${{ parameter.name|camelize }}
```
In the case where there is no type for some of the parameters, the result will look like
```php
public function someMethod(
    string $param1,
     $param2WithExtraneousWhitespace,
    int $param3
): ResponseInterface;
```

Please update this so that it doesn't have the additional whitespace.

I believe there are a couple other places that might have this same issue.

bayne commented on src/Command/GenerateSchemaCommand.php:
```diff
@@ -160,128 +179,170 @@ private function parseRequired($text)
+            if (preg_match($regexp, $description, $match)) {
+                foreach (['objectName', 'simple'] as $matchType) {
+                    if (isset($match[$matchType])) {
+                        // f**k https://core.telegram.org/bots/api#sendmediagroup
```
🤣 👍

bayne commented on src/Command/GenerateSchemaCommand.php:
```diff
@@ -57,98 +83,91 @@ protected function execute(InputInterface $input, OutputInterface $output)
+                'return'      => $this->getReturnType($description),
                 'description' => $description,
-                'link' => self::URL.$node->filter('a')->attr('href'),
+                'link'        => $node->filter('a')->attr('href'),
```
The generated client is currently setting all the links to `#setwebhook`. I ran into this issue before and was able to fix it with:

```suggestion
                'link'        => self::URL.$tableNode->previousAll()->filter('h4 a')->first()->attr('href'),
```

bayne commented on templates/ClientInterface.php.twig:
```diff
@@ -14,25 +13,20 @@ interface ClientInterface {
-{% elseif parameter.is_object and parameter.is_multiple_types == false %}
-        {{ parameter.type }} ${{ parameter.name|camelize }}{% if not parameter.required %} = null{% endif %}{% if loop.index < method.parameters|length %},{% endif %}
+    {% for parameter in method.parameters|sortByOptional %}
 
```
The generated client has an excess newline:

```php
    public function getUpdates(

        int $offset = null,
        int $limit = null,
        int $timeout = null,
        array $allowedUpdates = null
    ): ResponseInterface;
```

bayne reviewed (changes requested):
Thanks for you patience! 

I finally got around to testing this and noticed a couple things. Once those are addressed I'll do the merge.

maddmages-l2lnch committed:
PR review fix, strict types

madmages commented:
fixed

madmages commented:
kinda long responses.
I forked this library. I`ll continue work with it in my repo. https://github.com/madmages/TelegramApiGenerator
thanks for good idea ;)
